include/revng/ABIAnalyses/Common.h

```diff
@@ -0,0 +1,213 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/SmallSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/GlobalVariable.h"
+#include "llvm/IR/Instruction.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Value.h"
+#include "llvm/Support/Casting.h"
+
+#include "revng/BasicAnalyses/GeneratedCodeBasicInfo.h"
+#include "revng/MFP/MFP.h"
+#include "revng/Model/Binary.h"
+#include "revng/StackAnalysis/StackAnalysis.h"
+#include "revng/Support/revng.h"
+
+namespace ABIAnalyses {
+
+using llvm::BasicBlock;
+using llvm::CallInst;
+using llvm::GlobalVariable;
+using llvm::Instruction;
+using llvm::LoadInst;
+using llvm::SmallSet;
+using llvm::SmallVector;
+using llvm::StoreInst;
+using llvm::Value;
+
+using Register = model::Register::Values;
+
+enum TransferKind {
+  Read,
+  Write,
+  WeakWrite,
+  TheCall,
+  None,
+};
+
+struct ABIAnalysis {
+private:
+  SmallSet<GlobalVariable *, 20> ABIRegisters;
+  SmallVector<GlobalVariable *, 20> RegisterList;
+  const Instruction *CallSite;
+
+public:
+  ABIAnalysis(const GeneratedCodeBasicInfo &GCBI) :
+    ABIAnalysis(nullptr, GCBI){};
+
+  ABIAnalysis(const Instruction *CS, const GeneratedCodeBasicInfo &GCBI) :
+    RegisterList(), CallSite(CS) {
+
+    for (auto *CSV : GCBI.abiRegisters())
+      if (CSV) {
+        ABIRegisters.insert(CSV);
+        RegisterList.emplace_back(CSV);
+      }
+  };
+
+  const SmallVector<GlobalVariable *, 20> getRegisters() const {
+    return RegisterList;
+  }
+
+  bool isABIRegister(const Value *) const;
+
+  TransferKind classifyInstruction(const Instruction *) const;
+
+  SmallVector<const GlobalVariable *, 1>
+  getRegistersWritten(const Instruction *) const;
+
+  SmallVector<const GlobalVariable *, 1>
+  getRegistersRead(const Instruction *) const;
+};
+
+template<typename LatticeElementMap, typename CoreLattice, typename KeyT>
+typename CoreLattice::LatticeElement
+getOrDefault(const LatticeElementMap &S, const KeyT &K) {
+  auto V = S.find(K);
+  if (V != S.end()) {
+    return S.lookup(K);
+  }
+  return CoreLattice::DefaultLatticeElement;
+}
+
+template<typename LatticeElementMap, typename CoreLattice>
+LatticeElementMap
+combineValues(const LatticeElementMap &Lh, const LatticeElementMap &Rh) {
+
+  LatticeElementMap New = Lh;
+  for (const auto &[Reg, S] : Rh) {
+    New[Reg] = CoreLattice::
+      combineValues(getOrDefault<LatticeElementMap, CoreLattice>(New, Reg),
+                    getOrDefault<LatticeElementMap, CoreLattice>(Rh, Reg));
```

Comment:
```suggestion
    auto Lol = getOrDefault<LatticeElementMap, CoreLattice>(New, Reg);
    auto Lel = getOrDefault<LatticeElementMap, CoreLattice>(Rh, Reg);
    New[Reg] = CoreLattice::combineValues(Lol, Lel);
```
Same SLOCs, much more readable.